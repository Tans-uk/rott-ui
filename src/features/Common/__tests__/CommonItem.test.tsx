import {CommonItem, type CommonItemProps, type IconProps} from '@features'
import {formatMessage} from '@libs'
import {themeConfig} from '@providers'
import {fireEvent, render, waitFor} from '@utils'

describe('Common -> Common Item', () => {
  const onPressMock = jest.fn()
  const leftIconOnPressMock = jest.fn()
  const rightIconOnPressMock = jest.fn()

  const dummyData: CommonItemProps = {
    title: formatMessage('TEST.WITH.PARAM', {testText: 'TITLE'}),
    subTitle: formatMessage('TEST.WITH.PARAM', {testText: 'SUBTITLE'}),
    description: formatMessage('TEST.WITH.PARAM', {testText: 'DESCRIPTION'}),
    leftIcon: {
      name: 'STAR',
      width: 24,
      height: 24,
      color: themeConfig.colors.primary,
    },
    rightIcon: {
      name: 'ADD',
      width: 24,
      height: 24,
      color: themeConfig.colors.primary,
    },
  }

  const testId = {
    titleTestId: 'title-test-id',
    subTitleTestId: 'subtitle-test-id',
    descriptionTestId: 'description-test-id',
    leftIconTestId: 'left-icon-test-id',
    rightIconTestId: 'right-icon-test-id',
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('common item ilk renderlandığında snapshot ile eşleşmeli', () => {
    const commonItem = render(<CommonItem title={dummyData.title} />)

    expect(commonItem).toMatchSnapshot()
  })

  it('common item ilk renderlandığında title ekranda renderlanmali', () => {
    const {titleTestId} = testId
    const {title} = dummyData
    const {getByTestId} = render(<CommonItem title={title} />)

    const titleElement = getByTestId(titleTestId)

    expect(titleElement).toBeOnTheScreen()
  })

  it('common item ilk renderlandığında subTitle ekranda renderlanmali', () => {
    const {subTitleTestId} = testId
    const {title, subTitle} = dummyData
    const {getByTestId} = render(<CommonItem title={title} subTitle={subTitle} />)

    const subTitleElement = getByTestId(subTitleTestId)

    expect(subTitleElement).toBeOnTheScreen()
  })

  it('common item ilk renderlandığında description ekranda renderlanmali', () => {
    const {descriptionTestId} = testId
    const {title, subTitle, description} = dummyData
    const {getByTestId} = render(
      <CommonItem title={title} subTitle={subTitle} description={description} />
    )

    const descriptionElement = getByTestId(descriptionTestId)

    expect(descriptionElement).toBeOnTheScreen()
  })

  it('leftIcon ekranda renderlanmali', () => {
    const {leftIconTestId} = testId
    const {title, subTitle, description, leftIcon} = dummyData
    const {queryByTestId} = render(
      <CommonItem title={title} subTitle={subTitle} description={description} leftIcon={leftIcon} />
    )

    const leftIconElement = queryByTestId(leftIconTestId)

    expect(leftIconElement).toBeOnTheScreen()
  })

  it('leftIcon verilmediğinde ekranda renderlanmamali', () => {
    const {leftIconTestId} = testId
    const {title, subTitle, description} = dummyData
    const {queryByTestId} = render(
      <CommonItem title={title} subTitle={subTitle} description={description} />
    )

    const leftIconElement = queryByTestId(leftIconTestId)

    expect(leftIconElement).toBeNull()
  })

  it('leftIcon onPress özelliği verilmediğinde CommonItem onPress özelliğini tetiklemelidir.', async () => {
    const {leftIconTestId} = testId
    const {title, subTitle, description, leftIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        leftIcon={leftIcon}
        onPress={onPressMock}
      />
    )

    const leftIconElement = getByTestId(leftIconTestId)
    await waitFor(() => fireEvent.press(leftIconElement))

    expect(onPressMock).toHaveBeenCalled()
  })

  it('leftIcon onPress özelliği verildiğinde CommonItem onPress özelliğini tetiklenmemelidir.', async () => {
    const {leftIconTestId} = testId
    const {title, subTitle, description, leftIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        leftIcon={{onPress: leftIconOnPressMock, ...(leftIcon as IconProps)} as IconProps}
        onPress={onPressMock}
      />
    )

    const leftIconElement = getByTestId(leftIconTestId)
    await waitFor(() => fireEvent.press(leftIconElement))

    expect(leftIconOnPressMock).toHaveBeenCalled()
    expect(onPressMock).not.toHaveBeenCalled()
  })

  it('rightIcon ekranda renderlanmali', () => {
    const {rightIconTestId} = testId
    const {title, subTitle, description, rightIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        rightIcon={rightIcon}
      />
    )

    const rightIconElement = getByTestId(rightIconTestId)

    expect(rightIconElement).toBeOnTheScreen()
  })

  it('rightIcon verilmediğinde ekranda renderlanmamali', () => {
    const {rightIconTestId} = testId
    const {title, subTitle, description} = dummyData
    const {queryByTestId} = render(
      <CommonItem title={title} subTitle={subTitle} description={description} />
    )

    const rightIconElement = queryByTestId(rightIconTestId)

    expect(rightIconElement).toBeNull()
  })

  it('rightIcon onPress özelliği verilmediğinde CommonItem onPress özelliğini tetiklemelidir.', async () => {
    const {rightIconTestId} = testId
    const {title, subTitle, description, rightIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        rightIcon={rightIcon}
        onPress={onPressMock}
      />
    )

    const rightIconElement = getByTestId(rightIconTestId)
    await waitFor(() => fireEvent.press(rightIconElement))

    expect(onPressMock).toHaveBeenCalled()
  })

  it('rightIcon onPress özelliği verildiğinde CommonItem onPress özelliğini tetiklenmemelidir.', async () => {
    const {rightIconTestId} = testId
    const {title, subTitle, description, rightIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        rightIcon={{onPress: rightIconOnPressMock, ...(rightIcon as IconProps)} as IconProps}
        onPress={onPressMock}
      />
    )

    const rightIconElement = getByTestId(rightIconTestId)
    await waitFor(() => fireEvent.press(rightIconElement))

    expect(rightIconOnPressMock).toHaveBeenCalled()
    expect(onPressMock).not.toHaveBeenCalled()
  })

  it('common item ilk renderlangidinga left ve right iconu ekranda renderlanmali', () => {
    const {rightIconTestId, leftIconTestId} = testId
    const {title, subTitle, description, rightIcon, leftIcon} = dummyData
    const {getByTestId} = render(
      <CommonItem
        title={title}
        subTitle={subTitle}
        description={description}
        rightIcon={rightIcon}
        leftIcon={leftIcon}
      />
    )

    const rightIconElement = getByTestId(rightIconTestId)
    const leftIconElement = getByTestId(leftIconTestId)

    expect(rightIconElement).toBeOnTheScreen()
    expect(leftIconElement).toBeOnTheScreen()
  })

  it('common item ilk renderlandiginda left ve right iconu ekranda renderlanmamali', () => {
    const {rightIconTestId, leftIconTestId} = testId
    const {title, subTitle, description} = dummyData
    const {queryByTestId} = render(
      <CommonItem title={title} subTitle={subTitle} description={description} />
    )

    const rightIconElement = queryByTestId(rightIconTestId)
    const leftIconElement = queryByTestId(leftIconTestId)

    expect(rightIconElement).toBeNull()
    expect(leftIconElement).toBeNull()
  })
})
