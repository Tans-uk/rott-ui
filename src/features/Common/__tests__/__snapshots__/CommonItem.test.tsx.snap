// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common -> Common Item common item ilk renderlandığında snapshot ile eşleşmeli 1`] = `
<View
  flex={1}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      {
        "flex": undefined,
        "flexDirection": "column",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      animated={false}
      backgroundColor="grey-200"
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      overflowHidden={true}
      paddingHorizontal={16}
      paddingVertical={16}
      style={
        {
          "backgroundColor": "grey-200",
          "flex": 1,
          "flexDirection": "column",
          "overflow": "hidden",
          "paddingHorizontal": 31,
          "paddingVertical": 31,
        }
      }
    >
      <View
        alignItemsCenter={true}
        flex={1}
        justifyContentCenter={true}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          flex={1}
          style={
            {
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <Text
            size="full"
            style={
              {
                "color": "grey-900",
                "fontFamily": "Markpro-Medium",
                "fontSize": 16,
                "fontWeight": 600,
                "textAlign": "auto",
                "width": "100%",
              }
            }
            testID="title-test-id"
          >
            TEST.WITH.PARAM
          </Text>
          <View
            height={16}
            style={
              {
                "flex": undefined,
                "flexDirection": "column",
                "height": 31,
                "maxHeight": 31,
              }
            }
          >
            <View
              backgroundColor="rgba(234, 239, 240, 1)"
              borderRadius={4}
              height={16}
              overflowHidden={true}
              style={
                {
                  "backgroundColor": "rgba(234, 239, 240, 1)",
                  "flex": undefined,
                  "flexDirection": "column",
                  "height": 31,
                  "maxHeight": 31,
                  "overflow": "hidden",
                  "width": 429,
                }
              }
              width={223}
            >
              <View
                collapsable={false}
                jestAnimatedProps={
                  {
                    "value": {},
                  }
                }
                jestAnimatedStyle={
                  {
                    "value": {
                      "bottom": 0,
                      "flexDirection": "row",
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "transform": [
                        {
                          "translateX": -429,
                        },
                      ],
                    },
                  }
                }
                jestInlineStyle={{}}
                style={
                  [
                    {
                      "bottom": 0,
                      "flexDirection": "row",
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "transform": [
                        {
                          "translateX": -429,
                        },
                      ],
                    },
                  ]
                }
              >
                <BVLinearGradient
                  colors={
                    [
                      4293586928,
                      4294967295,
                      4293586928,
                    ]
                  }
                  endPoint={
                    {
                      "x": 1,
                      "y": 0,
                    }
                  }
                  locations={null}
                  startPoint={
                    {
                      "x": 0,
                      "y": 0,
                    }
                  }
                  style={
                    {
                      "height": 31,
                      "width": 429,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
</View>
`;
